File: board.py
# encoding=utf8

di = [0, 1, 0, -1]
dj = [1, 0, -1, 0]
ddi = [1, 1, -1, -1]
ddj = [1, -1, -1, 1]

def parsePosition(name):
  return (int(name[0]), int(name[1]))

def parseMove(move):
  return parsePosition(move[:2]) + parsePosition(move[2:])

def isRedPiece(piece):
  return piece.isupper()

def isKing(piece):
  return piece == 'k' or piece == 'K'

def inBoard(i, j):
  return i >= 0 and i < 10 and j >= 0 and j < 9

def inPalace(i, j):
  return ((i >= 0 and i <= 2) or (i >= 7 and i <= 9)) and j >= 3 and j <= 5

def inRedBase(i, j):
  return inBoard(i, j) and i >= 5

def inBlackBase(i, j):
  return inBoard(i, j) and i < 5

class board:
  def __init__(self):
    self.board = []
    for i in range(0, 10):
      self.board.append([])
      for j in range(0, 9):
        self.board[i].append("")
    self.redToGo = True
    self.moveHistory = []
    self.setBoard("rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR", True);

  def setBoard(self, fen, redToGo):
    rows = fen.split("/")
    for i in range(0, 10):
      j = 0
      for c in rows[i]:
        if c.isdigit():
          j += int(c)
        else:
          self.board[i][j] = c
          j += 1
    self.redToGo = redToGo

  def move(self, i1, j1, i2, j2):
    self.moveHistory.append((i1, j1, i2, j2, self.board[i2][j2]))
    self.board[i2][j2] = self.board[i1][j1]
    self.board[i1][j1] = ''
    self.redToGo = not self.redToGo

  def unmove(self):
    last = self.moveHistory.pop()
    self.board[last[0]][last[1]] = self.board[last[2]][last[3]]
    self.board[last[2]][last[3]] = last[4]
    self.redToGo = not self.redToGo

  def checkedMove(self, i1, j1, i2, j2):
    if self.board[i1][j1] == '' or isRedPiece(self.board[i1][j1]) != self.redToGo:
      return False
    moves = self.generatePieceMoves(self.board[i1][j1], i1, j1)
    if (i1, j1, i2, j2) not in moves:
      return False

    self.move(i1, j1, i2, j2)
    if self.hasWinningMove():
      self.unmove()
      return False
    return True

  def hasWinningMove(self):
    for move in self.generateAllMoves():
      if isKing(self.board[move[2]][move[3]]):
        return True
    return False

  def generateAllMoves(self):
    moves = []
    for i in range(0, 10):
      for j in range(0, 9):
        if self.board[i][j] != '' and isRedPiece(self.board[i][j]) == self.redToGo:
          moves += self.generatePieceMoves(self.board[i][j], i, j)
    return moves

  def isLosing(self):
    for move in self.generateAllMoves():
      if isKing(self.board[move[2]][move[3]]):
        return False
      self.move(move[0], move[1], move[2], move[3])
      notLosing = False
      if not self.hasWinningMove():
        notLosing = True
      self.unmove()
      if notLosing:
        return False
    return True

  def generatePieceMoves(self, piece, i, j):
    if piece.lower() == 'k':
      moves = self.generateKMoves(i, j)
    elif piece.lower() == 'a':
      moves = self.generateAMoves(i, j)
    elif piece.lower() == 'e':
      if isRedPiece(piece):
        moves = self.generateREMoves(i, j)
      else:
        moves = self.generateBEMoves(i, j)
    elif piece.lower() == 'h':
      moves = self.generateHMoves(i, j)
    elif piece.lower() == 'r':
      moves = self.generateRMoves(i, j)
    elif piece.lower() == 'c':
      moves = self.generateCMoves(i, j)
    elif piece.lower() == 'p':
      if isRedPiece(piece):
        moves = self.generateRPMoves(i, j)
      else:
        moves = self.generateBPMoves(i, j)

    filteredMoves = []
    for move in moves:
      target = self.board[move[2]][move[3]]
      if target == '' or isRedPiece(target) != self.redToGo:
        filteredMoves.append(move)
    return filteredMoves

  def generateKMoves(self, i, j):
    moves = []
    for r in range(0, 4):
      if inPalace(i + di[r], j + dj[r]):
        moves.append((i, j, i + di[r], j + dj[r]))
    for delta in [-1, 1]:
      ii = i
      while inBoard(ii + delta, j):
        ii += delta
        if isKing(self.board[ii][j]):
          moves.append((i, j, ii, j))
        elif self.board[ii][j] != '':
          break
    return moves

  def generateAMoves(self, i, j):
    moves = []
    for r in range(0, 4):
      if inPalace(i + ddi[r], j + ddj[r]):
        moves.append((i, j, i + ddi[r], j + ddj[r]))
    return moves

  def generateREMoves(self, i, j):
    moves = []
    for r in range(0, 4):
      if inRedBase(i + 2 * ddi[r], j + 2 * ddj[r]) and self.board[i + ddi[r]][j + ddj[r]] == '':
        moves.append((i, j, i + 2 * ddi[r], j + 2 * ddj[r]))
    return moves

  def generateBEMoves(self, i, j):
    moves = []
    for r in range(0, 4):
      if inBlackBase(i + 2 * ddi[r], j + 2 * ddj[r]) and self.board[i + ddi[r]][j + ddj[r]] == '':
        moves.append((i, j, i + 2 * ddi[r], j + 2 * ddj[r]))
    return moves

  def generateHMoves(self, i, j):
    table = [(1, 2, 0, 1), (1, -2, 0, -1), (-1, 2, 0, 1), (-1, -2, 0, -1),
             (2, 1, 1, 0), (2, -1, 1, 0), (-2, 1, -1, 0), (-2, -1, -1, 0)]
    moves = []
    for move in table:
      if inBoard(i + move[0], j + move[1]) and self.board[i + move[2]][j + move[3]] == '':
        moves.append((i, j, i + move[0], j + move[1]))
    return moves

  def generateRMoves(self, i, j):
    moves = []
    for r in range(0, 4):
      ni = i
      nj = j
      while inBoard(ni + di[r], nj + dj[r]):
        ni += di[r]
        nj += dj[r]
        moves.append((i, j, ni, nj))
        if self.board[ni][nj] != '':
          break
    return moves

  def generateCMoves(self, i, j):
    moves = []
    for r in range(0, 4):
      ni = i
      nj = j
      met = False
      while inBoard(ni + di[r], nj + dj[r]):
        ni += di[r]
        nj += dj[r]
        if self.board[ni][nj] != '':
          if met:
            moves.append((i, j, ni, nj))
            break
          else:
            met = True
        elif not met:
          moves.append((i, j, ni, nj))
    return moves

  def generateRPMoves(self, i, j):
    moves = []
    if inRedBase(i, j):
      moves.append((i, j, i - 1, j))
    else:
      if inBoard(i - 1, j):
        moves.append((i, j, i - 1, j))
      if inBoard(i, j - 1):
        moves.append((i, j, i, j - 1))
      if inBoard(i, j + 1):
        moves.append((i, j, i, j + 1))
    return moves

  def generateBPMoves(self, i, j):
    moves = []
    if inBlackBase(i, j):
      moves.append((i, j, i + 1, j))
    else:
      if inBoard(i + 1, j):
        moves.append((i, j, i + 1, j))
      if inBoard(i, j - 1):
        moves.append((i, j, i, j - 1))
      if inBoard(i, j + 1):
        moves.append((i, j, i, j + 1))
    return moves

  def output(self):
    for i in range(0, 10):
      print(self.board[i])

  def printMoves(self):
    moves = self.generateAllMoves()
    for move in moves:
      print('%d %d %d %d' % (move[0], move[1], move[2], move[3]))

File: internal/board.go
package internal

import (
	"fmt"
	"strings"
	"unicode"
)

type HistoryMove struct {
	Move    Move
	Capture string
}

type Board struct {
	Board       [10][9]string
	RedToGo     bool
	MoveHistory []HistoryMove
}

const (
	// The number of times the board position needs to be repeated in order to invoke repetition rules.
	REPETITION_COUNT = 3
)

func MakeInitialBoard() *Board {
	var board Board
	board.SetBoard("rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR", true)
	return &board
}

func (board *Board) SetBoard(fen string, redToGo bool) {
	rows := strings.Split(fen, "/")
	for i := 0; i < 10; i++ {
		for j := 0; j < 9; j++ {
			board.Board[i][j] = ""
		}
		j := 0
		for _, c := range rows[i] {
			if unicode.IsDigit(c) {
				j += int(c) - '0'
			} else {
				board.Board[i][j] = string(c)
				j += 1
			}
		}
	}
	board.RedToGo = redToGo
	board.MoveHistory = nil
}

func (board *Board) Print() {
	for _, row := range board.Board {
		for _, c := range row {
			if len(c) == 0 {
				fmt.Print(".")
			} else {
				fmt.Printf("%s", c)
			}
		}
		fmt.Println()
	}
}

func (board *Board) Move(move Move) {
	board.MoveHistory = append(board.MoveHistory, HistoryMove{
		Move:    move,
		Capture: *board.At(move.To),
	})
	*board.At(move.To) = *board.At(move.From)
	*board.At(move.From) = ""
	board.RedToGo = !board.RedToGo
}

func (board *Board) Unmove() {
	last := board.MoveHistory[len(board.MoveHistory)-1]
	board.MoveHistory = board.MoveHistory[:len(board.MoveHistory)-1]
	*board.At(last.Move.From) = *board.At(last.Move.To)
	*board.At(last.Move.To) = last.Capture
	board.RedToGo = !board.RedToGo
}

func (board *Board) CheckedMove(move Move) bool {
	if !inBoard(move.From) || !inBoard(move.To) {
		return false
	}

	if *board.At(move.From) == "" || isRedPiece(*board.At(move.From)) != board.RedToGo {
		return false
	}

	found := false
	for _, m := range board.generatePieceMoves(*board.At(move.From), move.From) {
		if m == move {
			found = true
			break
		}
	}
	if !found {
		return false
	}

	board.Move(move)
	if board.hasWinningMove() {
		board.Unmove()
		return false
	}

	return true
}

func (board *Board) IsLosing() bool {
	for _, move := range board.GenerateAllMoves() {
		if isKing(*board.At(move.To)) {
			return false
		}
		board.Move(move)
		notLosing := false
		if !board.hasWinningMove() {
			notLosing = true
		}
		board.Unmove()
		if notLosing {
			return false
		}
	}
	return true
}

func (board *Board) GenerateAllMoves() []Move {
	var moves []Move
	for i := 0; i < 10; i++ {
		for j := 0; j < 9; j++ {
			pos := Position{Row: i, Col: j}
			if *board.At(pos) != "" && isRedPiece(*board.At(pos)) == board.RedToGo {
				moves = append(moves, board.generatePieceMoves(*board.At(pos), pos)...)
			}
		}
	}
	return moves
}

func (board *Board) At(pos Position) *string {
	return &board.Board[pos.Row][pos.Col]
}

// Returns whether the same board position has repeated 3 times or more.
func (board *Board) CheckRepetition() bool {
	return board.checkRepetitionOfCycle(4) || board.checkRepetitionOfCycle(6) || board.checkRepetitionOfCycle(8)
}

var di = [...]int{0, 1, 0, -1}
var dj = [...]int{1, 0, -1, 0}
var ddi = [...]int{1, 1, -1, -1}
var ddj = [...]int{1, -1, -1, 1}

func isRedPiece(piece string) bool {
	return unicode.IsUpper(rune(piece[0]))
}

func isKing(piece string) bool {
	return piece == "k" || piece == "K"
}

func inBoard(pos Position) bool {
	return pos.Row >= 0 && pos.Row < 10 && pos.Col >= 0 && pos.Col < 9
}

func inPalace(pos Position) bool {
	return ((pos.Row >= 0 && pos.Row <= 2) || (pos.Row >= 7 && pos.Row <= 9)) && pos.Col >= 3 && pos.Col <= 5
}

func inRedBase(pos Position) bool {
	return inBoard(pos) && pos.Row >= 5
}

func inBlackBase(pos Position) bool {
	return inBoard(pos) && pos.Row < 5
}

func (board *Board) hasWinningMove() bool {
	for _, move := range board.GenerateAllMoves() {
		if isKing(*board.At(move.To)) {
			return true
		}
	}
	return false
}

func (board *Board) generatePieceMoves(piece string, pos Position) []Move {
	var moves []Move
	if strings.ToLower(piece) == "k" {
		moves = board.generateKMoves(pos)
	} else if strings.ToLower(piece) == "a" {
		moves = board.generateAMoves(pos)
	} else if strings.ToLower(piece) == "e" {
		if isRedPiece(piece) {
			moves = board.generateREMoves(pos)
		} else {
			moves = board.generateBEMoves(pos)
		}
	} else if strings.ToLower(piece) == "h" {
		moves = board.generateHMoves(pos)
	} else if strings.ToLower(piece) == "r" {
		moves = board.generateRMoves(pos)
	} else if strings.ToLower(piece) == "c" {
		moves = board.generateCMoves(pos)
	} else if strings.ToLower(piece) == "p" {
		if isRedPiece(piece) {
			moves = board.generateRPMoves(pos)
		} else {
			moves = board.generateBPMoves(pos)
		}
	}

	var filteredMoves []Move
	for _, move := range moves {
		target := *board.At(move.To)
		if target == "" || isRedPiece(target) != board.RedToGo {
			filteredMoves = append(filteredMoves, move)
		}
	}
	return filteredMoves
}

func generateShifts(pos Position, rowShift, colShift []int) []Move {
	var moves []Move
	for i := 0; i < len(rowShift); i++ {
		moves = append(moves, Move{
			From: pos,
			To: Position{
				Row: pos.Row + rowShift[i],
				Col: pos.Col + colShift[i],
			},
		})
	}
	return moves
}

func (board *Board) generateKMoves(pos Position) []Move {
	var moves []Move
	for _, move := range generateShifts(pos, di[:], dj[:]) {
		if inPalace(move.To) {
			moves = append(moves, move)
		}
	}
	for delta := -1; delta <= 1; delta += 2 {
		ii := pos.Row
		for ; ; ii += delta {
			to := Position{Row: ii + delta, Col: pos.Col}
			if !inBoard(to) {
				break
			}
			if isKing(*board.At(to)) {
				moves = append(moves, Move{From: pos, To: to})
			} else if *board.At(to) != "" {
				break
			}
		}
	}
	return moves
}

func (board *Board) generateAMoves(pos Position) []Move {
	var moves []Move
	for _, move := range generateShifts(pos, ddi[:], ddj[:]) {
		if inPalace(move.To) {
			moves = append(moves, move)
		}
	}
	return moves
}

func (board *Board) generateREMoves(pos Position) []Move {
	var moves []Move
	for r := 0; r < 4; r++ {
		to := Position{Row: pos.Row + 2*ddi[r], Col: pos.Col + 2*ddj[r]}
		check := Position{Row: pos.Row + ddi[r], Col: pos.Col + ddj[r]}
		if inRedBase(to) && *board.At(check) == "" {
			moves = append(moves, Move{From: pos, To: to})
		}
	}
	return moves
}

func (board *Board) generateBEMoves(pos Position) []Move {
	var moves []Move
	for r := 0; r < 4; r++ {
		to := Position{Row: pos.Row + 2*ddi[r], Col: pos.Col + 2*ddj[r]}
		check := Position{Row: pos.Row + ddi[r], Col: pos.Col + ddj[r]}
		if inBlackBase(to) && *board.At(check) == "" {
			moves = append(moves, Move{From: pos, To: to})
		}
	}
	return moves
}

func (board *Board) generateHMoves(pos Position) []Move {
	var table = [...][4]int{{1, 2, 0, 1}, {1, -2, 0, -1}, {-1, 2, 0, 1}, {-1, -2, 0, -1},
		{2, 1, 1, 0}, {2, -1, 1, 0}, {-2, 1, -1, 0}, {-2, -1, -1, 0}}
	var moves []Move
	for _, pattern := range table {
		to := Position{Row: pos.Row + pattern[0], Col: pos.Col + pattern[1]}
		check := Position{Row: pos.Row + pattern[2], Col: pos.Col + pattern[3]}
		if inBoard(to) && *board.At(check) == "" {
			moves = append(moves, Move{From: pos, To: to})
		}
	}
	return moves
}

func (board *Board) generateRMoves(pos Position) []Move {
	var moves []Move
	for r := 0; r < 4; r++ {
		ni := pos.Row
		nj := pos.Col
		for ; ; ni, nj = ni+di[r], nj+dj[r] {
			to := Position{Row: ni + di[r], Col: nj + dj[r]}
			if !inBoard(to) {
				break
			}
			moves = append(moves, Move{From: pos, To: to})
			if *board.At(to) != "" {
				break
			}
		}
	}
	return moves
}

func (board *Board) generateCMoves(pos Position) []Move {
	var moves []Move
	for r := 0; r < 4; r++ {
		ni := pos.Row
		nj := pos.Col
		met := false
		for ; ; ni, nj = ni+di[r], nj+dj[r] {
			to := Position{Row: ni + di[r], Col: nj + dj[r]}
			if !inBoard(to) {
				break
			}
			if *board.At(to) != "" {
				if met {
					moves = append(moves, Move{From: pos, To: to})
					break
				} else {
					met = true
				}
			} else if !met {
				moves = append(moves, Move{From: pos, To: to})
			}
		}
	}
	return moves
}

func (board *Board) generateRPMoves(pos Position) []Move {
	var moves []Move
	var rowShifts, colShifts []int
	if inRedBase(pos) {
		rowShifts = []int{-1}
		colShifts = []int{0}
	} else {
		rowShifts = []int{-1, 0, 0}
		colShifts = []int{0, 1, -1}
	}
	for _, move := range generateShifts(pos, rowShifts, colShifts) {
		if inBoard(move.To) {
			moves = append(moves, move)
		}
	}
	return moves
}

func (board *Board) generateBPMoves(pos Position) []Move {
	var moves []Move
	var rowShifts, colShifts []int
	if inBlackBase(pos) {
		rowShifts = []int{1}
		colShifts = []int{0}
	} else {
		rowShifts = []int{1, 0, 0}
		colShifts = []int{0, 1, -1}
	}
	for _, move := range generateShifts(pos, rowShifts, colShifts) {
		if inBoard(move.To) {
			moves = append(moves, move)
		}
	}
	return moves
}

func sameMoveSequence(seq1, seq2 []HistoryMove) bool {
	for i := 0; i < len(seq1); i++ {
		if seq1[i].Move != seq2[i].Move {
			return false
		}
	}
	return true
}

func isSamePlayerMoveSequenceCyclic(seq []HistoryMove) bool {
	for i := 0; i < len(seq); i++ {
		ni := (i + 1) % len(seq)
		if seq[i].Move.To != seq[ni].Move.From {
			return false
		}
	}
	return true
}

func isMoveSequenceCyclic(seq []HistoryMove) bool {
	extractPlayerMoves := func(start int) []HistoryMove {
		var newSeq []HistoryMove
		for i := start; i < len(seq); i += 2 {
			newSeq = append(newSeq, seq[i])
		}
		return newSeq
	}
	return isSamePlayerMoveSequenceCyclic(extractPlayerMoves(0)) &&
		isSamePlayerMoveSequenceCyclic(extractPlayerMoves(1))
}

func (board *Board) checkRepetitionOfCycle(cycle int) bool {
	moveCount := len(board.MoveHistory)
	lastCapture := -1
	for i := moveCount - 1; i >= 0; i-- {
		if len(board.MoveHistory[i].Capture) > 0 {
			lastCapture = i
			break
		}
	}
	recentNonCapturingMoves := moveCount - lastCapture - 1
	if recentNonCapturingMoves < cycle*REPETITION_COUNT {
		return false
	}

	for i := 1; i < REPETITION_COUNT; i++ {
		if !sameMoveSequence(board.MoveHistory[moveCount-cycle:moveCount],
			board.MoveHistory[moveCount-cycle*i-cycle:moveCount-cycle*i]) {
			return false
		}
	}

	return isMoveSequenceCyclic(board.MoveHistory[moveCount-cycle : moveCount])
}
